service-mirror.go

```diff
@@ -6,19 +6,330 @@ import (
 	"fmt"
 	"os"
 	"path/filepath"
+	"reflect"
 	"strings"
+	"time"
 
 	"github.com/sirupsen/logrus"
 	corev1 "k8s.io/api/core/v1"
 	apiError "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/watch"
 	"k8s.io/client-go/kubernetes"
-	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/cache"
+	client "k8s.io/client-go/tools/clientcmd"
 	"k8s.io/client-go/util/homedir"
 )
 
+// Make sure all global services lies in only one namespace.
+const GLOBAL_SVC_NAMESPACE = "default"
+
+func getServiceFilter(ctx Context) labels.Selector {
```

Comment:
Naming nit: we are not getting anything we are creating something. How about `mkWatchFilter` or `createWatcherFilter`? 

However, looking at this, maybe it should ideally be stored in a struct (along with other info we will need, e.g. client)?

```go
struct ServiceWatcher {
   client: k8s.API,
   labelSelector: labels.Selector
}

func (sw *ServiceWatcher) withServiceSelector(l labels.Selector) {
   sw.labelSelector = l
} 

func (sw *ServiceWatcher) WithDefaultSelector(l labels.Selector) {
    mirrorSvcLabel, err := labels.NewRequirement("mirror.linkerd.io/mirrored-service", selection.Equals, []string{"true"})
	if err != nil {
		ctx.log.Errorf("Unable to generate error requirement, Err : %v", err.Error())
	}

	// Create label requirements for "mirror.linkerd.io/headless-mirror-svc-name" not existing
	mirrorSvcParentLabel, err := labels.NewRequirement("mirror.linkerd.io/headless-mirror-svc-name", selection.DoesNotExist, []string{})
	if err != nil {
		ctx.log.Errorf("Unable to generate error requirement, Err : %v", err.Error())
	}

	// Create the label selector
	svcFilter := labels.NewSelector()
	svcFilter = svcFilter.Add(*mirrorSvcParentLabel)
	svcFilter = svcFilter.Add(*mirrorSvcLabel)
  
   sw.withSelector(svcFilter)
}
```

We could also offer a nicer API around this maybe and not make the `withDefaultSelector` public:

```go
func NewDefaultWatcher(client) ServiceWatcher {
   sw := &ServiceWatcher{client}
   sw.withDefaultSelector()
   sw
}

func main()
   client := makeClient()
  sw := sw.NewDefaultWatcher(client)
  sw.Run()
}
```

wdyt?

Reply:
+1 to this. This looks pretty nit, i will refactor my code :) and  as you have also give go ahead, I will do restructuring of code.

Reply:
Done added all fixes, main function, structs